lms/djangoapps/certificates/models.py

```diff
@@ -209,6 +212,8 @@ def get_queryset(self):
         """
         return super(EligibleCertificateManager, self).get_queryset().exclude(
             status__in=(CertificateStatuses.audit_passing, CertificateStatuses.audit_notpassing)
+        ).filter(
+            course_id__in=list(CourseOverview.objects.values_list('id', flat=True))
```

Comment:
Could this be written as `.exclude(course=None)`?

Reply:
unfortunately no - `course_id` on this model is a CourseKeyField as opposed to a foreign key, and the problem arises when a value in this field does not exist in the CourseOverview table.  For instance:

```
mysql> SELECT coc.id 'Course Overview Id', cg.course_id 'Certificate Course Id' FROM certificates_generatedcertificate cg LEFT JOIN course_overviews_courseoverview coc ON cg.course_id = coc.id WHERE cg.user_id = 2865145;
+---------------------+------------------------+
| Course Overview Id  | Certificate Course Id  |
+---------------------+------------------------+
| NULL                | MITx/16.101x/2013_SOND |
| MITx/16.110x/1T2014 | MITx/16.110x/1T2014    |
+---------------------+------------------------+
```

Reply:
OK. But `.exclude(course_id=None)` wouldn't work?

Reply:
no, for the same reason - the string value exists in `course_id` in the table being filtered here (GeneratedCertificate), so excluding where course_id=None still finds the problem cases like 'MITx/16.101x/2013_SOND' which has no corresponding row in CourseOverview.  The lack of a CourseOverview row is what causes the DoesNotExist exception

Reply:
Ah I get it now. :frowning_face: I suppose the surrounding comments are good enough pointers at this problem.